hazelcast/src/test/java/com/hazelcast/internal/partition/GracefulShutdownCorrectnessTest.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.partition;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.internal.partition.service.TestGetOperation;
+import com.hazelcast.internal.partition.service.TestPutOperation;
+import com.hazelcast.nio.Address;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.spi.OperationService;
+import com.hazelcast.test.HazelcastParametersRunnerFactory;
+import com.hazelcast.test.annotation.ParallelTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(Parameterized.class)
+@Parameterized.UseParametersRunnerFactory(HazelcastParametersRunnerFactory.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class GracefulShutdownCorrectnessTest extends PartitionCorrectnessTestSupport {
```

Comment:
Before merge:
```
11:10:02 [INFO] ------------------------------------------------------------------------
11:10:02 [INFO] Reactor Summary:
11:10:02 [INFO] 
11:10:02 [INFO] Hazelcast Root .................................... SUCCESS [0.510s]
11:10:02 [INFO] hazelcast ......................................... SUCCESS [16:58.865s]
11:10:02 [INFO] hazelcast-client .................................. SUCCESS [5:45.603s]
11:10:02 [INFO] hazelcast-spring .................................. SUCCESS [59.595s]
11:10:02 [INFO] hazelcast-build-utils ............................. SUCCESS [0.681s]
11:10:02 [INFO] ------------------------------------------------------------------------
```

After merge:
```
12:41:22 [INFO] ------------------------------------------------------------------------
12:41:22 [INFO] Reactor Summary:
12:41:22 [INFO] 
12:41:22 [INFO] Hazelcast Root .................................... SUCCESS [0.488s]
12:41:22 [INFO] hazelcast ......................................... SUCCESS [18:49.985s]
12:41:22 [INFO] hazelcast-client .................................. SUCCESS [7:56.258s]
12:41:22 [INFO] hazelcast-spring .................................. SUCCESS [1:04.291s]
12:41:22 [INFO] hazelcast-build-utils ............................. SUCCESS [0.681s]
12:41:22 [INFO] ------------------------------------------------------------------------
```

Effect on core module is 2mins. Client module build time seems increased 2mins too, I guess reason is client tests are using `hazelcast.shutdown()`.